<a href="https://colab.research.google.com/github/msantoshidevi/hotel_booking_project/blob/hotel_booking_analyses_modification/Hotel_booking_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    - Hotel Booking Analysis



##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Team Member**     - Santoshi Devi M

# **Project Summary -**

This exploratory data analysis(EDA) project focuses on analyzing the Hotel Booking dataset. This hotel booking dataset contain booking information for city hotel and resort hotel and includes information such as when the booking was made, length of stay, the number of adults, children and/or babies, and the number of available parking spaces, among other things.

It consists of 1,19,390 observations and 31 variables. Each observation represents a hotel booking. The dataset comprehense bookings arrive between the 1st of July of 2015 and 31st of August 2017, including bookings that effectively arrived and bookings that were cancelled.

Our analysis can help us understand about the given dataset and draw some useful insights to make some crucial business decisions regarding the factors affecting the bookings of the hotels. In hotel industry, cancellations and average daily rate are two important factors that effects the business.

By understanding the factors that are determining the cancellation of certain bookings, the hotels can take necessary precautions to reduce cancellation rate.

By understanding the patterns in average daily rate against different variables the hotels can be prepared in advance to generate more revenue and help to make more profitable business.

Our goal here is to understand such factors in the given data by performing exploratory data analysis. Exploratory Data Analysis refers to the critical process of performing initial investigations on data, so as to discover patterns, to spot anomalies to test hypothesis and to check assumptions with the help of summary statistics and graphical representations

The following are the various steps performed as a part of Exploratory Data Analysis:
* Data preparation
* Data cleaning
* Univariate analysis
* Bivariate analysis
* Multivariate analysis

**Data preparation:** Firstly, we imported libraries and dataset, some of the libraries used are NumPy, Pandas, matplotlib, seaborn. Once dataset and libraries are imported, process of analysis begins. But data have to be translated in an appropriate form. This process is known as data preparation.

**Data Cleaning:** The raw data received in the data received in the data set might not be directly suitable for analysis due to presence of unwanted data like, duplicate values, null values, outliers etc. We need to handle them first before we proceed with further analysis.

The duplicate values in the dataset to be removed, so that we get accurate patterns and numbers regarding the analysis.

The missing information for some or all features in some records, we need to either remove them or find alternatives to fill up the null values.

**Univariate Analysis:** Choose a single feature from the data and try to determine what the output or target value is, i.e. one features/variable at a time.
* Understand the trends and patterns of data.
* Analyze the frequency and other such characteristics of data.
* Know the distribution of the variables in the data.
* Visualise the relationship that may exist between different variables.

**Bivariate Analysis:** We try to analyze two features instead of one, and finally determine the classification of output we are looking for. It is a methodical statistical technique applied to a pair of variables of data to determine the empirical relationship between them. In  order words, it is meant to determine any concurrent relations.

**Multivariate Analysis:** We analyze three or more different features at a time, to understand the relationship between all the features involved.

After completing all the above processes, we have observed final conclusion from the patterns of the analysis and identified anomalies that can help increase the business objective. Based on findings, we recommand the hotels to implement necessary steps and plans to improve the marketing strategy.

# **GitHub Link -**

https://github.com/msantoshidevi/EDA_hotel_booking_project.git

# **Problem Statement**


We are here to explore a hotel booking dataset to discover important factors that govern the bookings, which contain booking information for a city hotel and resort hotel. We will analyze some important aspects of hotel bookings which will helps us identify major loopholes and give us insights which will be helpful to run profitable hotel business are as follows:
* Optimal length of stay to get the best daily rate?
* Which agent makes most number of bookings?
* which month results in high revenue?
* To predict whether or not a hotel was likely to receive a disproportionately high number of special requests?

#### **Define Your Business Objective?**

In Hotel industry, cancellation and Average Daily Rate  are two important factors that helps to run the business effectively.
- By understanding the factors that are determining the cancellation of a certain bookings, the hotels can take necessary precautions to reduce cancellation rate.
- By understanding the patterns in Average Daily Rate against different variables, the hotels can be prepared in advance to generate more revenue and help to make a profitable business.

Our goal here is to understand such factors in the given dataset by performing Exploratory Data Analysis.

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required.
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits.
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 20 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule.

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Dataset Loading

In [ ]:
# Load Dataset

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
hotel_df = pd.read_csv("/content/drive/MyDrive/Project/Hotel Bookings.csv")
hotel_df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

### Dataset First View

In [4]:
# Dataset First Look
hotel_df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

### Dataset Rows & Columns count

In [5]:
# Dataset Rows & Columns count
hotel_df.shape

(119390, 32)

**OBSERVATION**

* There are 119390 rows and 32 columns which means there are 119390 observations and 31 variables.



### Dataset Information

In [6]:
# Dataset Info
hotel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
# Creating a copy of dataframe
df = hotel_df.copy()

#### Duplicate Values

In [8]:
# Dataset Duplicate Value Count
df[hotel_df.duplicated()].shape

(31994, 32)

**OBSERVATION**
*   The number of duplicate rows are 31994 out of 119390.

In [9]:
df.drop_duplicates(inplace = True)

In [10]:
df.shape

(87396, 32)

**OBSERVATION**
*   After removing 31994 duplicate rows, we have 87396 rows and 32 columns in the dataset.

#### Missing Values/Null Values

In [11]:
# Missing Values/Null Values Count
df.isnull().sum().sort_values(ascending = False)[:6]

company               82137
agent                 12193
country                 452
children                  4
reserved_room_type        0
assigned_room_type        0
dtype: int64

In [12]:
df['agent'].unique()

array([ nan, 304., 240., 303.,  15., 241.,   8., 250., 115.,   5., 175.,
       134., 156., 243., 242.,   3., 105.,  40., 147., 306., 184.,  96.,
         2., 127.,  95., 146.,   9., 177.,   6., 143., 244., 149., 167.,
       300., 171., 305.,  67., 196., 152., 142., 261., 104.,  36.,  26.,
        29., 258., 110.,  71., 181.,  88., 251., 275.,  69., 248., 208.,
       256., 314., 126., 281., 273., 253., 185., 330., 334., 328., 326.,
       321., 324., 313.,  38., 155.,  68., 335., 308., 332.,  94., 348.,
       310., 339., 375.,  66., 327., 387., 298.,  91., 245., 385., 257.,
       393., 168., 405., 249., 315.,  75., 128., 307.,  11., 436.,   1.,
       201., 183., 223., 368., 336., 291., 464., 411., 481.,  10., 154.,
       468., 410., 390., 440., 495., 492., 493., 434.,  57., 531., 420.,
       483., 526., 472., 429.,  16., 446.,  34.,  78., 139., 252., 270.,
        47., 114., 301., 193., 182., 135., 350., 195., 352., 355., 159.,
       363., 384., 360., 331., 367.,  64., 406., 16

In [13]:
df['company'].unique()

array([ nan, 110., 113., 270., 178., 240., 154., 144., 307., 268.,  59.,
       204., 312., 318.,  94., 174., 274., 195., 223., 317., 281., 118.,
        53., 286.,  12.,  47., 324., 342., 373., 371., 383.,  86.,  82.,
       218.,  88.,  31., 397., 392., 405., 331., 367.,  20.,  83., 416.,
        51., 395., 102.,  34.,  84., 360., 394., 457., 382., 461., 478.,
       386., 112., 486., 421.,   9., 308., 135., 224., 504., 269., 356.,
       498., 390., 513., 203., 263., 477., 521., 169., 515., 445., 337.,
       251., 428., 292., 388., 130., 250., 355., 254., 543., 531., 528.,
        62., 120.,  42.,  81., 116., 530., 103.,  39.,  16.,  92.,  61.,
       501., 165., 291., 290.,  43., 325., 192., 108., 200., 465., 287.,
       297., 490., 482., 207., 282., 437., 225., 329., 272.,  28.,  77.,
       338.,  72., 246., 319., 146., 159., 380., 323., 511., 407., 278.,
        80., 403., 399.,  14., 137., 343., 346., 347., 349., 289., 351.,
       353.,  54.,  99., 358., 361., 362., 366., 37

**OBSERVATION**

*   Since, company and agent columns have company number and agent numbers as
data.
*   There may be some cases when customer didnt booked hotel via any agent or via any company.
*    So in that case values can be null under these columns.
*   We will replace null values by 0 in these columns

In [14]:
df[['company','agent']] = df[['company','agent']].fillna(0)

In [15]:
df['children'].unique()

array([ 0.,  1.,  2., 10.,  3., nan])

**OBSERVATION**

*  This column 'children' has 0 as value which means 0 children were present in group of customers who made that transaction.
*   So, 'nan' values are the missing values due to error of recording data.

In [16]:
df['children'].fillna(df['children'].mean(), inplace = True)

In [17]:
df['country'].unique()

array(['PRT', 'GBR', 'USA', 'ESP', 'IRL', 'FRA', nan, 'ROU', 'NOR', 'OMN',
       'ARG', 'POL', 'DEU', 'BEL', 'CHE', 'CN', 'GRC', 'ITA', 'NLD',
       'DNK', 'RUS', 'SWE', 'AUS', 'EST', 'CZE', 'BRA', 'FIN', 'MOZ',
       'BWA', 'LUX', 'SVN', 'ALB', 'IND', 'CHN', 'MEX', 'MAR', 'UKR',
       'SMR', 'LVA', 'PRI', 'SRB', 'CHL', 'AUT', 'BLR', 'LTU', 'TUR',
       'ZAF', 'AGO', 'ISR', 'CYM', 'ZMB', 'CPV', 'ZWE', 'DZA', 'KOR',
       'CRI', 'HUN', 'ARE', 'TUN', 'JAM', 'HRV', 'HKG', 'IRN', 'GEO',
       'AND', 'GIB', 'URY', 'JEY', 'CAF', 'CYP', 'COL', 'GGY', 'KWT',
       'NGA', 'MDV', 'VEN', 'SVK', 'FJI', 'KAZ', 'PAK', 'IDN', 'LBN',
       'PHL', 'SEN', 'SYC', 'AZE', 'BHR', 'NZL', 'THA', 'DOM', 'MKD',
       'MYS', 'ARM', 'JPN', 'LKA', 'CUB', 'CMR', 'BIH', 'MUS', 'COM',
       'SUR', 'UGA', 'BGR', 'CIV', 'JOR', 'SYR', 'SGP', 'BDI', 'SAU',
       'VNM', 'PLW', 'QAT', 'EGY', 'PER', 'MLT', 'MWI', 'ECU', 'MDG',
       'ISL', 'UZB', 'NPL', 'BHS', 'MAC', 'TGO', 'TWN', 'DJI', 'STP',
       'KNA', 'E

**OBSERVATION**
*   'country' column represents the country of origin of customer.
*   Since, this column has datatype of string. We will replace the missing value with the mode of 'country' column.

In [18]:
df['country'].fillna('others', inplace = True)

In [19]:
df.isnull().sum().sort_values(ascending = False)[:6]

hotel                          0
is_canceled                    0
reservation_status             0
total_of_special_requests      0
required_car_parking_spaces    0
adr                            0
dtype: int64

In [20]:
df[(df["adults"] == 0) & (df['children'] == 0) & (df['babies'] == 0)][["adults","children","babies","is_canceled"]]

adults  children  babies  is_canceled
2224         0       0.0       0            0
2409         0       0.0       0            0
3181         0       0.0       0            0
3684         0       0.0       0            0
3708         0       0.0       0            0
...        ...       ...     ...          ...
115029       0       0.0       0            0
115091       0       0.0       0            0
116251       0       0.0       0            0
116534       0       0.0       0            0
117087       0       0.0       0            0

[166 rows x 4 columns]

**OBSERVATION**

There are some rows with total number of adults, children or babies equal to zero. As adults, babies, and children can't be zero at the same time, so dropping the rows having all these zero at the same time because will mean that the hotel/room was not even booked.

In [21]:
df[df['adults']+df['babies']+df['children'] == 0].shape

(166, 32)

In [22]:
df.drop(df[df['adults']+df['babies']+df['children'] == 0].index, inplace = True)

In [23]:
df.shape

(87230, 32)

### What did you know about your dataset?

The dataset contains information about bookings and revenue for resots and hotels properties from July 2015 to August 2017. After filling missing values, removing duplicates and unwanted data removing, dataset consists of 87230 rows and 32 columns provides comprehensive details for analysis.

Our objective to analyze the data and gain insights into the best time for hotel bookings and the optimal length of stay to obtain the best rates. Additionally, we will explore other key aspects of the dataset to uncover valuable information.

By conducting a thorough analysis of this dataset, we aim to provide meaningful insights and recommendations to improve the booking strategies and revenue optimization for resort and hotel properties.